presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java

```diff
@@ -152,6 +160,91 @@ public static TupleDomainFilter toFilter(Domain domain, int notInThreshold)
         return MultiRange.of(rangeFilters, nullAllowed, false);
     }
 
+    /**
+     * Returns the exclusive number list if it could use NOT IN filter optimization.
+     *
+     * We're not able to distinguish NOT IN statement from normal range compare
+     * E.g.
+     * NOT IN (9, 10, 25) would be [Long.MIN_VALUE, 8] [11, 24] [26, Long.MAX_VALUE]
+     * (num < 8 OR (num > 11 AND number < 24) OR num > 26) could be the same
+     *
+     * When count of exclusive numbers is under the threshold, we can use NOT IN optimization
+     * it would return exclusive number list; otherwise empty list.
+     */
+    private static Optional<List<Long>> getExclusives(List<BigintRange> bigintRanges, int notInThreshold)
+    {
+        if (bigintRanges.size() < 2) {
+            return Optional.empty();
+        }
+
+        long count = getExclusiveCount(bigintRanges);
+        if (count >= notInThreshold || count < 0) {
+            return Optional.empty();
+        }
+
+        ImmutableList.Builder<Long> exclusiveNumbers = new ImmutableList.Builder<>();
+        long start = Long.MIN_VALUE;
+        for (int i = 0; i < bigintRanges.size(); i++) {
+            if (start < bigintRanges.get(i).getLower()) {
+                addExclusiveNumbers(exclusiveNumbers, start, bigintRanges.get(i).getLower() - 1);
+            }
+
+            if (bigintRanges.get(i).getUpper() == Long.MAX_VALUE) {
+                start = Long.MAX_VALUE;
+                break;
+            }
+
+            start = bigintRanges.get(i).getUpper() + 1;
+        }
+        if (start < Long.MAX_VALUE) {
+            addExclusiveNumbers(exclusiveNumbers, start, Long.MAX_VALUE);
+        }
+        return Optional.of(exclusiveNumbers.build());
+    }
+
+    private static void addExclusiveNumbers(ImmutableList.Builder<Long> exclusiveNumbers, long start, long end)
+    {
+        while (start <= end) {
+            exclusiveNumbers.add(start);
+            start++;
+        }
+    }
+
+    private static long getExclusiveCount(List<BigintRange> bigintRanges)
+    {
+        long count = 0;
+        long start = Long.MIN_VALUE;
+        for (int i = 0; i < bigintRanges.size(); i++) {
+            if (start < bigintRanges.get(i).getLower()) {
+                // Handle out of boundary cases
+                if (bigintRanges.get(i).getLower() - start < 0) {
```

Comment:
this condition is never true

Reply:
> this condition is never true

So in this condition I was trying to handle the case when we have the result that exceed the range of LONG

For example:
if the `bigintRanges.get(i).getLower()` is 10, our `start` is Long.MIN_VALUE, then `bigintRanges.get(i).getLower() - start` would overflow thus becoming negative, do you think this is valid?

Reply:
> if the bigintRanges.get(i).getLower() is 10, our start is Long.MIN_VALUE, then bigintRanges.get(i).getLower() - start would overflow thus becoming negative
If you remove this if check, the code is the same. The later check
```
if (count < 0) {
                    return -1;
                }
```
Would return for this case anyways.

Reply:
@kewang1024 I actually think you can tighten the condition even more by passing the `notInThreshold` to `getExclusiveCount` function. In the loop in `getExclusiveCount`, whenever the number of distinct values exceeds this threshold, you can end the loop early.

Reply:
> > if the bigintRanges.get(i).getLower() is 10, our start is Long.MIN_VALUE, then bigintRanges.get(i).getLower() - start would overflow thus becoming negative
> 
> If you remove this if check, the code is the same. The later check
> 
> ```
> if (count < 0) {
>                     return -1;
>                 }
> ```
> 
> Would return for this case anyways.

Thanks! That's a nice catch